### ISSUE TITLE
Number of batches returned correct?

### ISSUE THREAD

**Issue by sha168** [tristandeleu/pytorch-meta#67]
Hi,
First of all, thank you for sharing the data loaders. I have come across a strange behaviour and was wondering if this is the intended behaviour. When sampling batches from the datasets, the sampler does not seem to sample from all images and the number of batches returned is limited to these images.

For example, when sampling from the Omniglot dataset, with 1028 classes and 5 num_classes_per_task, the number of batches returned is equal to C(1028, 5) and seems limited to only 5 images per class. Similarly, when setting num_classes_per_task=1, we get C(1028,1)=1028 batches, which again consist of the same image for each class (with shuffle=True).

Here is a small example. For the num_classes_per_task=1 we see that the number of unique classes that are being sampled is around 647 while the total number of classes is 1028, which is fine as we are randomly sampling. However, the number of unique images sampled is also 647, which indicates that we are really just sampling from one image per class (1028 images in total). Am I missing something?

```
import random
import torch.optim
from torchvision.transforms import Compose, ToTensor, Resize
from torchmeta.transforms import ClassSplitter, Categorical
from torchmeta.datasets import Omniglot
from torchmeta.utils.data import BatchMetaDataLoader
import numpy as np
from collections import Counter

seed = 0
random.seed(seed)
torch.manual_seed(seed)
torch.cuda.manual_seed_all(seed)

print("num_classes_per_task=1")
omniglot_dataset = Omniglot("data",
                   num_classes_per_task=1,
                   transform=Compose([Resize(28), ToTensor()]),
                   meta_train=True,
                   download=True)
omniglot_dataset = ClassSplitter(omniglot_dataset, shuffle=True, num_train_per_class=1)
omniglot_dataloader = BatchMetaDataLoader(omniglot_dataset, batch_size=1, num_workers=1)

batch_list = []
target_list = []
for batch in omniglot_dataloader:
    batch_list.append(np.asarray(batch["train"][0]))
    target_list.append(batch["train"][1][0][0][0])

print("Number of batches: ", len(omniglot_dataloader))
print("Number of unique classes: ", len(Counter(target_list).keys()))
print("Number of unique images: ",np.unique(np.stack(batch_list), axis=0).shape[0])

print("*****")
print("num_classes_per_task=5")
omniglot_dataset = Omniglot("data",
                   num_classes_per_task=5,
                   transform=Compose([Resize(28), ToTensor()]),
                   meta_train=True,
                   download=True)
omniglot_dataset = ClassSplitter(omniglot_dataset, shuffle=True, num_train_per_class=5)
omniglot_dataloader = BatchMetaDataLoader(omniglot_dataset, batch_size=1, num_workers=1)
print("Number of batches: ", len(omniglot_dataloader))
```

**Comment by tristandeleu** [comment]
Hi! Thank you for the kind words!

The reason why you only get 647 unique classes here instead of 1028 is due to the way the sampler inside the dataloader creates batches of episodes, when `shuffle=True` in `BatchMetaDataLoader` (which is the default value).
https://github.com/tristandeleu/pytorch-meta/blob/91daf7ff8a4883c58609942f229282af967ead88/torchmeta/utils/data/sampler.py#L40-L44

Here the sampler `CombinationRandomSampler` samples tuples of size `num_classes_per_task` **with replacement**. The correct way to do it would be to sample tuples of size `num_classes_per_task` without replacement. However in general there is a combinatorially large amount of such tuples (i.e. of possible tasks), so it would be impossible to samples without replacement; in practice, when you have such a large population, the probability of sampling the same tuple twice is so infinitesimally small that it is effectively behaving like sampling without replacement. Here the population is so small (1028) that you can clearly see the effect of replacement.

You can verify that this is working as expected when setting `shuffle=False` in `BatchMetaDataLoader`.
```python
# ...
omniglot_dataloader = BatchMetaDataLoader(omniglot_dataset, batch_size=1, shuffle=False, num_workers=1)
# ...
# num_classes_per_task=1
# Number of batches:  1028
# Number of unique classes:  1028
# Number of unique images:  1028
```

Now when you create a task (indexed by a tuple), the images sampled to create the train and test datasets will always be the same, regardless of the value of `shuffle` in `ClassSplitter`, to ensure reproducibility. There is a one-to-one correspondance between a task and a dataset. Here is a small example, when you can see this behavior when you call the same task multiple times:

```python
task1 = omniglot_dataset[(0,)]
img1, label1 = task1['train'][0]
del task1

task2 = omniglot_dataset[(0,)]
img2, label2 = task2['train'][0]
assert torch.all(img1 == img2)
```

What `shuffle=True` in `ClassSplitter` does is shuffle the images within the same class when creating the episode, so that you don't get in a situation where, for example, all the tasks that share the same class (say `images_background/Alphabet_of_the_Magi/character01`) always have the same image (if `num_train_per_class=1`).

```python
omniglot_dataset = Omniglot("data",
                   num_classes_per_task=2,
                   transform=Compose([Resize(28), ToTensor()]),
                   meta_train=True,
                   download=True)
omniglot_dataset = ClassSplitter(omniglot_dataset, shuffle=True, num_train_per_class=1)

task1 = omniglot_dataset[(0, 1)]
img1, label1 = task1['train'][1]
del task1

task2 = omniglot_dataset[(1, 2)]
img2, label2 = task2['train'][0]
assert label1 == label2
assert not torch.all(img1 == img2)
```

I hope this helps!
